src/libs/ComposerFocusManager.ts

```diff
@@ -1,25 +1,273 @@
         return;
     }
-    resolveIsReadyToFocus();
+    focusMap.set(id, {input, businessType});
+    if (!input) {
+        return;
+    }
+    input.blur();
 }
 
-function isReadyToFocus(): Promise<void> {
-    return isReadyToFocusPromise;
+/**
+ * On web platform, if we intentionally click on another input box, there is no need to restore focus.
+ * Additionally, if we are closing the RHP, we can ignore the focused input.
+ */
+function focus(input: InputElement, shouldIgnoreFocused = false) {
+    if (!input) {
+        return;
+    }
+    if (shouldIgnoreFocused) {
+        isWindowReadyToFocus().then(() => input.focus());
+        return;
+    }
+    const activeInput = getActiveInput();
+    if (activeInput) {
+        return;
+    }
+    isWindowReadyToFocus().then(() => input.focus());
 }
 
+/**
+ * Restore the focus state after the modal is dismissed.
+ */
+function restoreFocusState(
+    id: ModalId,
+    shouldIgnoreFocused = false,
+    restoreFocusType: RestoreFocusType = CONST.MODAL.RESTORE_FOCUS_TYPE.DEFAULT,
+    businessType: BusinessType = CONST.MODAL.BUSINESS_TYPE.DEFAULT,
+) {
+    if (!id || !activeModals.length) {
+        return;
+    }
+    const index = activeModals.indexOf(id);
+
+    // This id has been removed from the stack.
+    if (index < 0) {
+        return;
+    }
+    activeModals.splice(index, 1);
+    if (restoreFocusType === CONST.MODAL.RESTORE_FOCUS_TYPE.PRESERVE) {
+        return;
+    }
+
+    const {input} = focusMap.get(id) ?? {};
+    focusMap.delete(id);
+    if (restoreFocusType === CONST.MODAL.RESTORE_FOCUS_TYPE.DELETE) {
+        return;
+    }
+
+    // This modal is not the topmost one, do not restore it.
+    if (activeModals.length > index) {
+        if (input) {
+            const lastId = activeModals.slice(-1)[0];
+            focusMap.set(lastId, {...focusMap.get(lastId), input});
+        }
+        return;
+    }
+    if (input) {
+        focus(input, shouldIgnoreFocused);
+        return;
+    }
+
+    // Try to find the topmost one and restore it
+    const stack = [...focusMap].filter(([, v]) => v.input && v.businessType === businessType);
+    if (stack.length < 1) {
+        return;
+    }
+    const [lastId, {input: lastInput}] = stack.slice(-1)[0];
+
+    // The previous modal is still active
+    if (activeModals.indexOf(lastId) >= 0) {
+        return;
+    }
+    focus(lastInput, shouldIgnoreFocused);
+    focusMap.delete(lastId);
+}
+
+function resetReadyToFocus(id: ModalId) {
+    const promise: PromiseMapValue = {
+        ready: Promise.resolve(),
+        resolve: () => {},
+    };
+    promise.ready = new Promise<void>((resolve) => {
+        promise.resolve = resolve;
+    });
+    promiseMap.set(id, promise);
+}
+
+/**
+ * Backward compatibility, for cases without an ID, it's fine to just take the topmost one.
+ */
+function getKey(id: ModalId) {
+    if (id) {
+        return id;
+    }
```

Comment:
```suggestion
function getKey() {
```

lets handle this case where we use it, also the name is not clear.